text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+That still doesn't protect unpublished packages using unstable stdlib crates without reserved names from breakage, but due to their use of unstable interfaces we have no obligation to keep them working.
+Also, once we have an option to explicitly provide the source for stdlib deps, they can force the behavior they want.
+This seems good enough.
+
+
+# Drawbacks
+
+ - The mock registry for sysroot binaries is a disgusting hack.
+
+ - Only some crates in the rust repo (at least `core`, `alloc` and `collections`) can properly be built just based upon their `Cargo.toml`.
+   However it's precisely only these "foundational" crates that will be of interest to freestanding developers.
+   Hosted developers can likely get pre-built binaries for the platform they need with `rustup`, just as they do today.
+
+ - Compilers could provide crates in their sysroot that don't match the Rust specification, and Cargo would be none the wiser.
+   (Technically, this problem already exists with falling back on the sysroot binaries, but users will probably expect better when they can specify standard library dependencies explicitly.)
+   Since the *interface* of the stdlib is specified, it would be neat if we could put a big crate type/interface on crates.io, which compiler implementations would need to match.
+
+
+# Alternatives
+
+ - Instead of copying binaries from the sysroot, we could just leave it to rustc to find them.
+   But then a simple `--no-resolve-sysroot` would not work, and the logic for passing `--extern` would need be more complicated.
+
+ - Previous versions of this RFC were a simpler but more brittle.
+   Please refer to the git history to see them.
+
+# Unresolved questions
+
+ - Users of the stable compiler should be able to build the stdlib from source, since it is trusted, but cannot because it uses unstable features.
+   Some notion of a trusted package/registry or way to route the secret bootstrap key would be required to fix this.
+
+ - It is unclear what should go in the lockfile when building with sysroot binaries.
+
+ - Whether to add the `--no-resolve-sysroot` flag to rustc, as described above.
+
+ - Should `cargo new` specify `std`, or any other stdlib crates explicitly by default?
+   I'd hope so!
+
+ - Should one be able to opt-out of implicit build and development dependencies?
+   Currently, it makes sense to always make `std` available for `build.rs` since it must exist for the compiler.
+   But if platform-specific parts of the `std` are exposed only with features or "scenarios" (a newly-proposed mechanism specifically for handling environment differences), then we lose an opportunity to be able to express mandatory cross-compiling.
+   Finally, in the far future it may be possible to build rustc on platforms where all of `std` isn't available, invalidating the reasoning that `std` is never unavailable as a build dependency.
```

Comment:
Perhaps this would be better phrased as "where not all of `std` is available" - @Ericson2314, the current phrasing reads as "where none of `std` is available.

My understanding is that such platforms already exist - for example, there's work on Rust on Cortex M4 microcontrollers, which lacks some parts of `std`, and not all of `std` really makes sense in Robigalia's Phoma environment either as far as I can tell.